ethcore/sync/src/chain/mod.rs

```diff
@@ -675,6 +684,55 @@ pub struct ChainSync {
     warp_sync: WarpSync,
 }
 
+#[derive(Debug, Default)]
+struct GetPooledTransactionsReport {
+    found: H256FastSet,
+    missing: H256FastSet,
+    not_sent: H256FastSet,
+}
+
+impl GetPooledTransactionsReport {
+    fn generate(
+        mut asked: Vec<H256>,
+        received: impl IntoIterator<Item = H256>,
+    ) -> Result<Self, H256> {
+        let mut out = GetPooledTransactionsReport::default();
+
+        let asked_set = asked.iter().copied().collect::<H256FastSet>();
+        let mut asked_iter = asked.drain(std::ops::RangeFull);
+        let mut txs = received.into_iter();
+        let mut next_received: Option<H256> = None;
+        loop {
+            if let Some(received) = next_received {
+                if !asked_set.contains(&received) {
+                    return Err(received);
+                }
+
+                if let Some(requested) = asked_iter.next() {
```

Comment:
I guess we assume that both iterators have to be sorted, right? I'm not sure I understand the point of the entire loop then, if this again could be done with simple (and efficient) set operations.
We basically have two sets: `received` and `asked` and we need:
```
superfluous = received - asked;
found = received * asked;
missing = asked - found;
```

Reply:
Okay, seems we also want to split `missing` into `requested-but-missing` and `not_sent`. But that should be doable as well (given all the resulting sets are ordered: `not_sent = asked[last(found)...], missing = asked[...last(found)] - found`).

If you prefer a loop I also feel this could be simplified if we keep the collections ordered (unless it's not possible cause `H256` doesn't do `PartialOrd` - don't remember exactly).
```
let received : VecDeque<>;
let asked: Iterator<H256>; // ordered by `H256`
loop {
   match (asked.next(), received.pop_front()) {
     (Some(expected), Some(got)) if expected == got => {
          found.insert(expected);
     },
     (Some(expected), Some(got)) if expected < got => {
          missing.insert(expected);
          received.push_front(got);
     },
     (Some(expected), Some(got)) if expected > got => {
          return Err(got);
     },
     (None, Some(got)) => {
          return Err(got);
     },
     (Some(expected), None) => {
         not_sent.insert(expected); // we can early-exit here and just append the rest of the iterator.
     },
     (None, None) => break;
   }
}
return Ok(out);
```